### ISSUE TITLE
Mode 2 runs into harvester

### ISSUE THREAD

**Issue by abehnke** [Courseplay/courseplay#2456]
During beet harvesting, I get occassional "is in traffic" messages.  There appear to be two situations: (1) when the tipper approaches the harvester+tractor from rear and turns toward the pipe and (2) at the end of the row after unloading, the tipper turns into the harvester tractor rather than away. Equipment: Grimme Rootster 604, Case IH 1455 XL, Kröger Agroliner HKD 302.  

Standard map: Goldcrest Valley, Field 11
### Courseplay: initialized 48/48 files (v5.02.00027)
[log.txt](https://github.com/Courseplay/courseplay/files/1479664/log.txt)
![fsscreen_2017_11_16_12_32_47xx](https://user-images.githubusercontent.com/3346683/32907780-3fdde5f6-cacf-11e7-9094-c9f488d3e045.jpg)
![fsscreen_2017_11_16_12_28_18xx](https://user-images.githubusercontent.com/3346683/32907784-436c2d2c-cacf-11e7-9f22-153f97250001.jpg)

**Comment by pops64** [comment]
Do you have AIVE installed? I think this is problem with the bybassing mode cutting in front of the harvester which it shouldn't do. I have seen it once

**Comment by abehnke** [comment]
Thanks for looking into this.  I do not have AIVE installed.  I have managed to find a repeatable situation for the case where the tipper approaches from behind.  The setup is that the tipper is parked to the left of the harvester+tractor.  Thus it heads toward the harvester+tractor and circles around back.  Unfortunately, it has started the turn too close to the harvester+tractor and does not have enough space to straighten out to be parallel to the harvester+tractor. I am experimenting with mode2.lua - STATE_DRIVE_TO_COMBINE - to see if I can have a better setup for this movement.

**Comment by pops64** [comment]
It is a problem with leaving the combine. It should be driving away from the combine instead of cutting in front of it. Which is why I think it is a bypassing problem which overrides where it is told to go if it has a traffic conflict. I believe debug channel 4 prints mode 2 please post a log that active

**Comment by abehnke** [comment]
Thanks for the quick response.   First, some other general info.  I have path finding turned off so the tipper may drive on the beets.  The field is large and the harvester and tipper tanks are small so there can be some waiting at 100%. 

I think I have a work around for my situation. I made my changes under a predicate that tests for the combine.cp.isGrimmeRootster604 so this is not a general fix but it works for me!

For the case where the tipper approaches the stopped tractor+harvester from the left.  I made the movements similar to that of the combine.cp.isChopper.  The idea is to approach directly from the rear then when the pipe state starts it can easily move right a bit with out any risk of collision.

For the case where the tipper has unloaded at the end of the row.  Rather than make a 3 part turn to the left (collision with the tractor+harvester) - I make a 3 part turn to the right.  This is done by setting the first WP to slightly right rear of the tipper.  When the WP is executed, tipper makes a 180 turn to the right.

I also found an occassional problem that when the tractor+harvester is empty at the end of a row, the tipper the goes straight ahead and parks - this is OK, but then the tractor+harvester decides to turn.  When it makes the turn it does a slight wiggle to the right before turning left thus another "in traffic" but this time from the tractor+harvester against the tipper. The fix was similar to the one above.  I have attached a diff file of my changes to mode2.lua for your amusement.

Tell me if you still want a log file.
[patch-111717.txt](https://github.com/Courseplay/courseplay/files/1483251/patch-111717.txt)

**Comment by pops64** [comment]
That is why asked for the log file because the waypoints for the tipper turning away are based of the offset set. I wanted to see if CP was setting them correctly

**Comment by abehnke** [comment]
Ok - attached is a log file with a run that generates the two "is in traffic" messages.  I removed my changes from the lua,  so I am running a pure zip file of v5.02.00027 for this log file.  Some other comments that may be of use:
1. I have no problem with my other harvester - Case IH 1660 Axial-Flow - using it for many days now with no warnings.
2. Most of the beet field is processed by the Grimme Rootster 604 with no "is in traffic" messages.
3. The pipe is centered well over the tipper when using the 'auto' defaults.
4. Before changing the mode2.lua code, I tried various values in the HUD offsets but nothing seem to work.
5. I know about issue #1331 "Tipper blocks combi during harvest" from Jun 11 2016 - This may still be true but not for what I am doing now.

Thanks for looking into this! Tell me if you need some further info.
[log.txt](https://github.com/Courseplay/courseplay/files/1484023/log.txt)

**Comment by pops64** [comment]
Alright could you know do a separate one with the traffic debug channel 3. Thanks It is performing the math correctly now I want to see if it is encountering a traffic collision.

**Comment by abehnke** [comment]
Here is the log with channel 3 turned on.  Same conditions as my prior log run.  The first collision with the grimmeRootster604 is the turn from the left to do an unload.  I released this stop by pressing Ctrl + num 7 twice  - I don't have to move the tipper.  (It changes its mind about the collision situation after I do the Ctrl + num 7 twice. Interesting?)   The second collision with caseIH1455XL (the harvester puller tractor) is at the end of the row upon an unload.  These two situations are outlined in my first post. 

[log.txt](https://github.com/Courseplay/courseplay/files/1484601/log.txt)

**Comment by pops64** [comment]
Yea has I figured there is traffic collision on the tipper leaving and it is entering bypassing mode which is probably causing this. I'll see if I can reproduce this if not could you provide your save game and courses you are using to produce this thanks

**Comment by abehnke** [comment]
Here are the links for the save game and courses.  The tipper uses "f11 unload" and the tractor+harvester uses "f11 3m".

[savegame3.zip](https://github.com/Courseplay/courseplay/files/1484834/savegame3.zip)
[CoursePlay_Courses.zip](https://github.com/Courseplay/courseplay/files/1484839/CoursePlay_Courses.zip)

**Comment by Blackmankot** [comment]
Hello! I want to add from myself that the same problem is observed with grain combines. And when the transport is fully loaded, he does not wait for the combine to leave, but immediately begins the movement trying to get ahead of the combine and a collision occurs.

**Comment by Jimmy-the-Geek** [comment]
Same here. I was harvesting a gigantic field of potatoes, and more often than not the tipper would turn into the harvester to go unload..

**Comment by lcc1211** [comment]
I concur with @Blackmankot on grain combines. It seems to happen every time the tractor heads toward the stopped combine.  Example: The tractor is heading west towards the stopped combine which is facing east.  The tractor passes the combine and turns toward the pipe but either ends up blocked behind the combine or is stopped because the combine header is in the way.

**Comment by JLConradie** [comment]
Hello
I agree with Blackmankot. It happens not only with potato harvesters but with grain harvesters as well.
It is really a problem making courseplay unusable to harvest fields. Its driving me up the wall.
Collisions every time the support tractor is leaving the combine. It just turns into the the harvester every time if the tractor/overloder is full. It does not matter were I place my starting and ending points of the dumping route.  as soon as the tractor leaves the combine.  The problem is getting worse with every update of the game or courseplay.
Sorry that I step on toes but its really is a HUGE problem. Everything is working like a charm until that point. Don't change anything to the chopper that section is working very well. 

Regards

**Comment by pvajko** [comment]
@pops64 I don't think it is a bypass problem, I looked at the savegeme from @abehnke, what I see is that it sets up the course to the waiting position incorrectly as the tractor is in the fruit, the course briefly appears before the collision is detected, see my draft:

Honestly, I'm not sure that this specific situation ever worked but who knows. 

Can someone post a savegame and courses for a setup where the tipper is not in the fruit?

![2456](https://user-images.githubusercontent.com/2379521/33071347-cd0b207e-ce80-11e7-92e4-abcf1bd21c0a.jpg)

**Comment by abehnke** [comment]
In reply to pvajko - Yes, I setup the tractor+tipper on the wrong side of the field.  I usually don't do that but I forgot the rule. I had just bought the field of beets and was excited to get it harvested.  The depot is right at the field so I thought this will be a quick drop off operation.  Sorry about that.  With my other fields for the grain harvester I follow the rule plus I have path finding turned on.  When I am paying attention I try and keep the tipper out of the fruit.  But it think the second problem (the end of the row situation) remains since I don't believe the unload course is involved with that movement.

**Comment by pvajko** [comment]
I know, that's why we need a savegame with tractor not on the fruit side, to reproduce that end of the row issue.

**Comment by Blackmankot** [comment]
I suggest - to increase the indentation of the automatic transport from the combine by 20%. And crash stops at the entrance to the combine are less frequent. At the same time, unloading is performed.
In this screenshot, the tractor indentation is selected for the auto course. And the tractor is next to the header.
![2017-11-21_23-56-45](https://user-images.githubusercontent.com/26131714/33082855-c414084e-cf18-11e7-96cd-2f053ba95108.jpg)
And in this case I increased the indent manually. Unloading is performed! And the tractor less often avarino stops.
![2017-11-21_23-48-43](https://user-images.githubusercontent.com/26131714/33083402-3c94d068-cf1a-11e7-832b-dea1d309c3ba.jpg)

**Comment by pops64** [comment]
@pvajko yea if the unloading course is on the other side of the tipper when it reaches 100

**Comment by abehnke** [comment]
I moved the unload course to the non-fruit side of field 11.  The results are the same for the "in traffic" at the end of the row.  The first approach with an empty tipper was successful but if I let it run after awhile there came a time when the tipper was parked in the non-fruit area but the harvester's pipe was on the fruit side so the tipper had to approach from the left, go behind with the result of an "in traffic" message.   When testing, I start the tipper first and the tractor+harvester 2nd. The tipper course is "f11 unload NF" and the tractor+harvester is "f11 3m".  Thanks for looking into this.  Tell me if you need anything else.


[log.txt](https://github.com/Courseplay/courseplay/files/1492135/log.txt)
[savegame3.zip](https://github.com/Courseplay/courseplay/files/1492137/savegame3.zip)
[CoursePlay_Courses.zip](https://github.com/Courseplay/courseplay/files/1492138/CoursePlay_Courses.zip)

**Comment by JLConradie** [comment]
Hello....
Just follow these steps. After A LOT of testing This works every time. Hopefully this simple drawing will explain.  

The starting point of the dump rout is VERY important !!! 

=> THE STARTING POINT  MUST BE AT THE LOWER BACK OF THE COMBINE IF COMBINE IS MOVING WITH PIPE IN CROPS.

After I have done this no collisions with combine after filling Trailer or support vehicle. They all turn away from the Combine. It does not matter what the situation is. Believe me it works. This is how Courseplay is generating the course and three point turn from the combine.

Hopefully it will help someone. I know that not all the maps or fields are easy.


![no collition](https://user-images.githubusercontent.com/32715611/33088200-a542e6f6-cef5-11e7-91aa-9dcab44925ee.png)

**Comment by lcc1211** [comment]
@JLConradie I have tried starting my route from every direction and if the combine fills up with the pipe on the fruit side - the tractor will end up in traffic unless it is already behind the combine.  Shape of the field and starting point seem to have no affect.  I've reproduced this from all 4 corners of fields 1, 13, 8, 9, 10, 16 and 14 on Hobbs Farm, on 15, 16, 17, 18, 19 on Westbridge Hills and several other maps.

**Comment by pvajko** [comment]
@lcc1211 so if you reproduced it on so many maps and fields, why not post a savegame and courses so we can also reproduce a situation like this as shown in #2463?

https://user-images.githubusercontent.com/11431031/32979977-e183e7bc-cc5e-11e7-826b-2d6f3ea964fc.jpg

**Comment by pops64** [comment]
Starting point only affects this if the havester is at 100. The code for it to wait on the havester is skipped in this case. The end point only matters to combine detection. This traffic collision is eithier the parking manvuer being spullied with incorrect waypoints or something overriding it's normal behavior.

**Comment by lcc1211** [comment]
I have that one as well as at the end of the field. I'll get you the saved game.

**Comment by lcc1211** [comment]
You can use the standard TC590 and its default header. Just set the header's width to 5.4.

**Comment by JLConradie** [comment]
@pops64 well As you can see its a problem. What I have noticed is that when I upgraded from 1.4.4.0 to 1.5.1.0 with the same Courseplay version the support tractor's behaviour got worse. It started to turn into the combine instead of the other way.  It turn right instead of left. Maybe its in the game and not courseplay. Could be, I don't know.

**Comment by pvajko** [comment]
@lcc1211 great, we need now is the savegame and courses with vanilla equipment to reproduce it.

**Comment by abehnke** [comment]
I got version 5.02.00011 from ModHub and ran my non-fruit side course on field 11. (same test as the savegame3 in my post of yesterday). Success! The end of row turn of the tipper is away from the tractor+harvester so no "in traffic".  I did find a third kind of "in traffic" problem  that happens in both that latest and this older version.  In this situation, the tractor+harvester is "in traffic" against the tipper.  It happens  near the end of the row.  The unload has happened and tipper drives forward and parks - the tractor+harvester does not move at this time.  Unfortunately, it decides to park where the tractor+harvester wants to do part of the wiggle of its turn around movement.  Maybe this a unique combination of this kind of harvester and the "L" shape of field 11 since it appears to happen on the shorter side (east) of the "L" shape? This only happened twice most of the movements are either the 3 part turn-around a the end of the row or the wait and pull forward movement for the middle of the row.

**Comment by pops64** [comment]
Thanks.So we now know it is probably a problem with v18 or v20. We are looking into this.

**Comment by ThomasGaertner** [comment]
please try v.0034 or higher and report....

**Comment by lcc1211** [comment]
@ThomasGaertner 
[log.txt](https://github.com/Courseplay/courseplay/files/1497655/log.txt)
[SaveGame.zip](https://github.com/Courseplay/courseplay/files/1497658/SaveGame.zip)

v.0034 did a little bit different. 
![20171122175152_1](https://user-images.githubusercontent.com/24764870/33153085-02eee63e-cfae-11e7-8e28-c4a7d34215ed.jpg)

**Comment by JLConradie** [comment]
@ThomasGaertner
Hello Thomas...
I have tested it.. You are spot on.. It seems working like it should.  Both scenarios.. are fixed.
1) On the end of the field.  Tractor / Support vehicle not full Combine empty -- FIXED Tractor turns to the left and not into combine any more. 
2) Pipe in crops Combine not empty but tractor/Support vehicle full... FIXED Tractor turns to the left and not into combine any more.
Thank You Thank You!!

You know Thomas. I'm soooo happy.  I wish I could send you the best bottle of wine from the Western Cape in South Africa were I'm staying!!...

Now I can enjoy the game once more and play it the way it should be played.  
Regards Louis

**Comment by pops64** [comment]
Closing this has fixed. @lcc1211 the picture you posted is a completely separate issue. You can either bump up the horizontal offset to give the tractor more room. Decrease your field speed, Or just wait until the harvester moves further down field has this will fix itself.

**Comment by JLConradie** [comment]
Thank You Thomas. I'll do that.

**Comment by ThomasGaertner** [comment]
@JLConradie : https://www.paypal.com/us/cgi-bin/webscr?cmd=_flow&SESSION=sQSzGODuKJ1Rg3CRcpMmtzgXnZQ_LaBW832K9mbSJW25cHbKUCLOmyto1ja&dispatch=5885d80a13c0db1f8e263663d3faee8dc3f308debf7330dd8d0b0a9f21afd7d3&rapidsState=Donation__DonationFlow___StateDonationLogin&rapidsStateSignature=b25260be2a03cae6acec90fbe2f3bb1086fd45b6
:-)
